[ctx_prof] CtxProfAnalysis: populate module data

mtrofin opened the pull request:
Continuing from #102084, which introduced the analysis, we now populate it with info about functions contained in the module.

When we will update the profile due to e.g. inlined callsites, we'll ingest the callee's counters and callsites to the caller. We'll move those to the caller's respective index space (counter and callers), so we need to know and maintain where those currently end.

We also don't need to keep profiles not pertinent to this module.

This patch also introduces an arguably much simpler way to track the GUID of a function from the frontend compilation, through ThinLTO, and into the post-thinlink compilation step, which doesn't rely on keeping names around. A separate RFC and patches will discuss extending this to the current PGO (instrumented and sampled) and other consumers as an infrastructural component.

Issue #89287

mtrofin committed:
[ctx_prof] CtxProfAnalysis: populate module data

Continuing from #102084, which introduced the analysis, we now populate
it with info about functions contained in the module.

When we will update the profile due to e.g. inlined callsites, we'll
ingest the callee's counters and callsites to the caller. We'll move
those to the caller's respective index space (counter and callers), so
we need to know and maintain where those currently end.

We also don't need to keep profiles not pertinent to this module.

This patch also introduces an arguably much simpler way to track the GUID
of a function from the frontend compilation, through ThinLTO, and into
the post-thinlink compilation step, which doesn't rely on keeping names
around. A separate RFC and patches will discuss extending this to the
current PGO (instrumented and sampled) and other consumers as an
infrastructural component.

snehasish commented on llvm/include/llvm/Analysis/CtxProfAnalysis.h:
```diff
@@ -9,23 +9,38 @@
-public:
-  explicit PGOContextualProfile(PGOCtxProfContext::CallTargetMapTy &&Profiles)
-      : Profiles(std::move(Profiles)) {}
+  GlobalValue::GUID getKnownGUID(const Function &F) const;
```
Prefer `getGUID(...)` and document it returns zero if unknown and also define what it means to be `known`?

mtrofin replied:
why - there's a `GlobalValue::getGUID`, at least `getKnownGUID` tells you it's a bit different. Added the comment... also renamed to `getDefinedFunctionGUID`, it's more clear I think.

snehasish commented on llvm/include/llvm/Transforms/Instrumentation/PGOCtxProfLowering.h:
```diff
@@ -24,5 +24,26 @@ class PGOCtxProfLoweringPass : public PassInfoMixin<PGOCtxProfLoweringPass> {
+/// pre-computing and assigning as metadata, this mechanism is resilient to such
+/// changes (as well as name changes e.g. suffix ".llvm." additions). It's
+/// arguably a much simpler mechanism than PGO's current GV-based one, and can
+/// be made available more broadly.
```
I don't think you need to make a case for the guid computation in the comments here. I suggest dropping the contents from "It's arguably .....". The comment will be stale in the future.

mtrofin replied:
It should also change at that point because this whole thing would move out :) But removed that bit.

snehasish commented on llvm/lib/Analysis/CtxProfAnalysis.cpp:
```diff
@@ -81,7 +83,55 @@ CtxProfAnalysis::Result CtxProfAnalysis::run(Module &M,
+    if (F.isDeclaration())
+      continue;
+    auto GUID = AssignUniqueIDPass::getGUID(F);
+    assert(GUID);
```
Maybe `assert(GUID && "Guid not found ....");` otherwise this just looks like `assert(0)` in the failure message.

mtrofin replied:
done

snehasish commented on llvm/lib/Transforms/Instrumentation/PGOCtxProfLowering.cpp:
```diff
@@ -349,3 +350,34 @@ bool CtxInstrumentationLowerer::lowerFunction(Function &F) {
+    return GlobalValue::getGUID(F.getGlobalIdentifier());
+  }
+  auto *MD = F.getMetadata(GUIDMetadataName);
+  assert(MD);
```
assert(MD && "missing guid metadata");

mtrofin replied:
done

snehasish commented on llvm/test/Analysis/CtxProfAnalysis/full-cycle.ll:
```diff
@@ -0,0 +1,110 @@
+;
+; RUN: llvm-ctxprof-util fromJSON --input=%t/profile.json --output=%t/profile.ctxprofdata
+;
+; Disable pre-inline to avoid losing the trivial functions to the preinliner.
```
This is just a limitation of the test right? If so can we avoid this by adding noinline attributes instead?

mtrofin replied:
yes. either would work

snehasish replied:
Can you change it use the attribute, I'm a little wary of disabling critical passes we rely on in tests.

mtrofin committed:
comments

snehasish commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1196,6 +1196,7 @@ PassBuilder::buildModuleSimplificationPipeline(OptimizationLevel Level,
     // In pre-link, we just want the instrumented IR. We use the contextual
     // profile in the post-thinlink phase.
     // The instrumentation will be removed in post-thinlink after IPO.
+    MPM.addPass(AssignUniqueIDPass());
```
Should there be a TODO here to move this AssignUniqueIDPass earlier in the pipeline in the future?

mtrofin replied:
done. Come to think of it, it may eventually make sense for this to be an analysis, but this is for a RFC to decide. Left the

snehasish commented on llvm/lib/Analysis/CtxProfAnalysis.cpp:
```diff
@@ -81,7 +83,56 @@ CtxProfAnalysis::Result CtxProfAnalysis::run(Module &M,
+  // .Profiles.
+  // Trim first the roots that aren't in this module.
+  DenseSet<GlobalValue::GUID> ProfiledGUIDs;
+  for (auto &[RootGuid, Tree] : llvm::make_early_inc_range(*MaybeCtx))
```
Do you need to check MaybeCtx before deref? 

Also Tree is unused so maybe just `for auto&[RootGuid, Unused]`?

mtrofin replied:
MaybeCtx is checked up at line 81.

snehasish commented on llvm/include/llvm/Analysis/CtxProfAnalysis.h:
```diff
@@ -35,6 +51,20 @@ class PGOContextualProfile {
+  }
+
+  uint32_t allocateNextCounterIndex(const Function &F) {
+    assert(isFunctionKnown(F));
+    return FuncInfo.find(getDefinedFunctionGUID(F))->second.NextCounterIndex++;
```
Is this performing the lookup 2x on an assert build? If so maybe rewrite this as 
```
auto& [found, iter] = FuncInfo.find(....);
(void) found;
assert(found && "function not found");
```

mtrofin replied:
ya, but I wanted the code to be more readable - i.e. the assertion is that the function is known rather than its implementation.

snehasish commented on llvm/include/llvm/Analysis/CtxProfAnalysis.h:
```diff
@@ -9,23 +9,39 @@
 #ifndef LLVM_ANALYSIS_CTXPROFANALYSIS_H
 #define LLVM_ANALYSIS_CTXPROFANALYSIS_H
 
+#include "llvm/ADT/StringMap.h"
```
I didn't see a StringMap being introduced, there is a new DenseMap though?

mtrofin replied:
+1

snehasish reviewed (approved):
lgtm, with some minor comments. The only one I have a strong opinion on is the use of attr vs disable-preinline for the test.

mtrofin committed:
addressed comments

mtrofin committed:
Merge branch 'main' into ctx_analysis

snehasish reviewed (approved):
lgtm

mtrofin committed:
moved AssignUniqueIDPass to Analysis so that llvm-link can build (also makes more sense)

mtrofin committed:
Make tests resilient to test path

AssignUniqueIDPass skips over functions with assigned ID. We can manually set
the ID for local linkage functions in test. Otherwise, as their GUID is computed
in a way that incorporates the module name (i.e. path), the tests would depend
on the build path.

snehasish commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -47,6 +47,7 @@ MODULE_PASS("always-inline", AlwaysInlinerPass())
 MODULE_PASS("annotation2metadata", Annotation2MetadataPass())
 MODULE_PASS("attributor", AttributorPass())
 MODULE_PASS("attributor-light", AttributorLightPass())
+MODULE_PASS("assign-guid", AssignUniqueIDPass())
```
Looks like the surrounding items are sorted by alphabetical order, so this should come after "annotation2metadata"?

snehasish commented on llvm/lib/Analysis/CtxProfAnalysis.cpp:
```diff
@@ -64,10 +66,40 @@ Value toJSON(const PGOCtxProfContext::CallTargetMapTy &P) {
 } // namespace json
 } // namespace llvm
 
+const char *AssignUniqueIDPass::GUIDMetadataName = "unique_id";
```
Missed this earlier, why not just `guid`?

snehasish commented on llvm/lib/Analysis/CtxProfAnalysis.cpp:
```diff
@@ -66,6 +66,36 @@ Value toJSON(const PGOCtxProfContext::CallTargetMapTy &P) {
+  for (auto &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+    if (F.getMetadata(GUIDMetadataName))
```
Document the behaviour that we only assign a new guid if it does not exist?

snehasish reviewed (approved):
lgtm

mtrofin committed:
comments: mostly renamed stuff, also setting `source_filename` in test to keep
guids stable

mtrofin committed:
Merge branch 'main' into ctx_analysis
